stdlib/Random/src/XoshiroSimd.jl

```diff
@@ -0,0 +1,344 @@
+        while i+8 <= len
+            resLoc = _and(res, 0x0101010101010101)
+            res = _lshr(res, 1)
+            unsafe_store!(reinterpret(Ptr{UInt64}, dst + i), resLoc)
+            i += 8
+        end
+        resLoc = _and(res, 0x0101010101010101)
+        ref = Ref(resLoc)
+        GC.@preserve ref ccall(:memcpy, Nothing, (Ptr{UInt8}, Ptr{Cvoid}, Csize_t), dst+i, pointer_from_objref(ref) , len-i)
+        t = _shl17(s1)
+        s2 = _xor(s2, s0)
+        s3 = _xor(s3, s1)
+        s1 = _xor(s1, s2)
+        s0 = _xor(s0, s3)
+        s2 = _xor(s2, t)
+        s3 = _rotl45(s3)
+    end
+    if rng isa TaskLocal
+        task.rngState0, task.rngState1, task.rngState2, task.rngState3 = s0, s1, s2, s3
+    else
+        rng.s0, rng.s1, rng.s2, rng.s3 =  s0, s1, s2, s3
+    end
+    nothing
+end
+
+
+@noinline function xoshiro_bulk_simd(rng::Union{TaskLocal, Xoshiro}, dst::Ptr{UInt8}, len::Int, ::Type{T}, ::Val{N}) where {T,N}
+    # we rely on specialization, llvm const-prop + instcombine to remove unneeded masking
+    mskOR =  oneBits(T, Val(N))
+    mskAND = mskBits(T, Val(N))
+
+    s0, s1, s2, s3 = forkRand(rng, Val(N))
+
+    i = 0
+    while i + 8*N <= len
+        res = _rotl23(_plus(s0,s3))
+        res = _or(_and(res, mskAND), mskOR)
+        t = _shl17(s1)
+        s2 = _xor(s2, s0)
+        s3 = _xor(s3, s1)
+        s1 = _xor(s1, s2)
+        s0 = _xor(s0, s3)
+        s2 = _xor(s2, t)
+        s3 = _rotl45(s3)
+        unsafe_store!(reinterpret(Ptr{NTuple{N,VecElement{UInt64}}}, dst + i), res)
+        i +=  8*N
+    end
+    return i
+end
+
+@noinline function xoshiro_bulk_simd(rng::Union{TaskLocal, Xoshiro}, dst::Ptr{UInt8}, len::Int, ::Type{Bool}, ::Val{N}) where {N}
+    s0, s1, s2, s3 = forkRand(rng, Val(N))
+    msk = ntuple(i->VecElement(0x0101010101010101), Val(N))
+    i = 0
+    while i + 64*N <= len
+        res = _rotl23(_plus(s0,s3))
+        t = _shl17(s1)
+        s2 = _xor(s2, s0)
+        s3 = _xor(s3, s1)
+        s1 = _xor(s1, s2)
+        s0 = _xor(s0, s3)
+        s2 = _xor(s2, t)
+        s3 = _rotl45(s3)
+        for k=0:7
+            tmp = _lshr(res, k)
+            toWrite = _and(tmp, msk)
+            unsafe_store!(reinterpret(Ptr{NTuple{N,VecElement{UInt64}}}, dst + i + k*N*8), toWrite)
+        end
+        i +=  64*N
+    end
+    return i
+end
+
+
+function rand!(rng::Union{TaskLocal, Xoshiro}, dst::Union{Array{Float32}, UnsafeView{Float32}}, ::SamplerTrivial{CloseOpen12{Float32}})
+    GC.@preserve dst xoshiro_bulk(rng, convert(Ptr{UInt8}, pointer(dst)), length(dst)*4, Float32, xoshiroWidth())
+    @inbounds @simd for i = 1:length(dst)
+        dst[i] = dst[i] - 1.0f0
+    end
+    dst
+end
+
+function rand!(rng::Union{TaskLocal, Xoshiro}, dst::Union{Array{Float64}, UnsafeView{Float64}}, ::SamplerTrivial{CloseOpen12{Float64}})
+    GC.@preserve dst xoshiro_bulk(rng, convert(Ptr{UInt8}, pointer(dst)), length(dst)*8, Float64, xoshiroWidth())
+    @inbounds @simd for i = 1:length(dst)
```

Comment:
Can we add a single pass method?
We could make `xoshiro_bulk_simd` except a function argument to apply to `toWrite` before storing it, which can do the masking to `Float(64/32)` and subtract 1.0.

For big arrays, passing over the data once could be a big performance win.